File: leetcode/dynamic_programming/combination_sum4/js/solution.js
function solve(nums, target) {
  const dp = Array(target + 1).fill(0);
  dp[0] = 1;
  for (let i = 1; i <= target; i++) {
    for (const num of nums) {
      if (i - num >= 0) dp[i] += dp[i - num];
    }
  }
  return dp[target];
}

if (require.main === module) {
  console.log(solve([1, 2, 3], 4));
}

module.exports = { solve };

File: leetcode/dynamic_programming/combination_sum4/scala/Solution.scala
object Solution {
  def solve(nums: Array[Int], target: Int): Int = {
    val dp = Array.fill(target + 1)(0)
    dp(0) = 1
    for (i <- 1 to target) {
      for (num <- nums) {
        if (i - num >= 0) dp(i) += dp(i - num)
      }
    }
    dp(target)
  }
}

object SolutionMain {
  def main(args: Array[String]): Unit = {
    println(Solution.solve(Array(1, 2, 3), 4))
  }
}
